File: Programmers/Lv.1/개인정보 수집 유효기간/150370.java
import java.util.*;

class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        int[] answer = {};
        
        Map<String, Integer> termsMap = new HashMap<>();
        for (String term : terms) {
            String[] t = term.split(" ");
            termsMap.put(t[0], Integer.parseInt(t[1]));
        }
        
        List<Integer> list = new ArrayList<>();
        
        for (int i = 0; i < privacies.length; i++) {
            String[] privacy = privacies[i].split(" ");
            String pdate = privacy[0];
            int pterm = termsMap.get(privacy[1]);
            
            int pyy = Integer.parseInt(pdate.substring(0, 4));
            int pmm = Integer.parseInt(pdate.substring(5, 7)) + pterm;
            int pdd = Integer.parseInt(pdate.substring(8)) - 1;
            
            if (pdd == 0) {
                pmm -= 1;
                pdd = 28;
            }
            
            if (pmm > 12) {
                pyy += pmm / 12;
                pmm %= 12;
                
                if (pmm == 0) {
                    pyy -= 1;
                    pmm = 12;
                }
            }
            
            int yy = Integer.parseInt(today.substring(0, 4));
            int mm = Integer.parseInt(today.substring(5, 7));
            int dd = Integer.parseInt(today.substring(8));
            
            if (yy > pyy || (yy == pyy && mm > pmm) || (yy == pyy && mm == pmm && dd > pdd)) {
                list.add(i + 1);
            }
        }
        
        answer = new int[list.size()];
        for (int j = 0; j < list.size(); j++) {
            answer[j] = list.get(j);
        }
        
        return answer;
    }
}

File: Programmers/Lv.1/개인정보 수집 유효기간/150370.js
function solution(today, terms, privacies) {
    const answer = [];
    
    const termsObj = {};
    for (const term of terms) {
        const [t1, t2] = term.split(" ");
        termsObj[t1] = parseInt(t2);
    }
        
    for (let i = 0; i < privacies.length; i++) {
        const [pdate, pterm] = privacies[i].split(" ");
        
        let pyy = parseInt(pdate.slice(0, 4));
        let pmm = parseInt(pdate.slice(5, 7)) + termsObj[pterm];
        let pdd = parseInt(pdate.slice(8)) - 1;
        
        if (pdd == 0) {
            pmm -= 1;
            pdd = 28;
        }
        
        if (pmm > 12) {
            pyy += parseInt(pmm / 12);
            pmm %= 12;
            
            if (pmm == 0) {
                pyy -= 1;
                pmm = 12;
            }
        }
        
        const yy = parseInt(today.slice(0, 4));
        const mm = parseInt(today.slice(5, 7));
        const dd = parseInt(today.slice(8));
        
        if (yy > pyy || (yy == pyy && mm > pmm) || (yy == pyy && mm == pmm && dd > pdd)) {
            answer.push(i + 1);
        }
    }
    
    return answer;
}
